### PR TITLE
Scrape out monkey patching

### PR DESCRIPTION
**Pull request by pirj**
Sibling PRs:
 - https://github.com/rspec/rspec-expectations/pull/1245
 - https://github.com/rspec/rspec-mocks/pull/1365
 - https://github.com/rspec/rspec-support/pull/462

Prerequisite PRs:
 - https://github.com/rspec/rspec-rails/pull/2420
 - https://github.com/rspec/rspec-support/pull/457
 - https://github.com/rspec/rspec-rails/pull/2425

- [ ] Remove temporary commit. CI is expected to fail on Core/Mocks/Support/Rails steps, as `disable_monkey_patching!` calls methods that were removed with no replacement. I'll use the `maintenance-branch` temporary commit hack to fraternize those sibling PRs.

 http://rspec.info/blog/2013/07/the-plan-for-rspec-3/#zero_monkey_patching_mode

 > we do want to encourage people to switch to the new syntax, so we plan to make RSpec 3 print a warning on first usage of any the old syntax methods (should, should_not, should_receive, etc) unless the should syntax has been explicitly enabled. This should nudge folks towards the new syntax while keeping RSpec friendly to new users and will pave the way for the old syntax to be disabled by default in RSpec 4.

 > zero-monkey-patching mode for RSpec...  We plan for these config options to become the defaults in RSpec 4.0, so that RSpec 4.0 will have zero monkey patching out of the box.

 As for "disabled by default" vs "completely removed" and "default, out of the box" vs "impossible" I can only say that RSpec 4 was probably planned to be released earlier, as:

 > we'll probably be dropping support for 1.8.7 in RSpec 4

 but we've also dropped 1.9, 2.0, 2.1 and 2.2

 https://github.com/rspec/rspec-core/issues/2301#issuecomment-234612382

 > In RSpec 4, we plan to extract all monkey patching from RSpec and move it into a separate gem, so that monkey patching is opt-in instead of opt-out and users have to explicitly install and load a gem to get it.

 `rspec-should` (or `rspec-monkey` as it's also about exposing example group DSL in the top-level/Module?) will be released later.

 Those using the monkey-patched `should` syntax are not encouraged to update to RSpec 4 until this gem is extracted.

 Those using the globally-exposed DSL are encouraged to use `RSpec.describe`/`RSpec.shared_examples_for` instead.

### CODE REVIEW

File: spec/rspec/core/configuration_spec.rb [hunk 1]
```diff
@@ -2375,20 +2373,13 @@ def metadata_hash(*args)
...
         end
 
-        Module.class_exec do
-          def my_group_method; :original; end
-        end
-
         config.alias_example_group_to :my_group_method
 
         expect(ExampleGroup.my_group_method).to be < ExampleGroup
-        expect(Module.new.my_group_method).to be < ExampleGroup
```
Comment (JonRowe) [inline spec/rspec/core/configuration_spec.rb:2391]: What was this changed?

File: spec/rspec/core/configuration_spec.rb [hunk 2]
```diff
@@ -312,12 +312,10 @@ def mod.configuration; @config ||= Struct.new(:custom_setting).new; end
       it_behaves_like "a configurable framework adapter", :mock_with
 
       it "allows rspec-mocks to be configured with a provided block" do
-        mod = Module.new
-
-        expect(RSpec::Mocks.configuration).to receive(:add_stub_and_should_receive_to).with(mod)
+        expect(RSpec::Mocks.configuration).to receive(:verify_partial_doubles=).with(true)
 
         config.mock_with :rspec do |c|
-          c.add_stub_and_should_receive_to mod
```
Comment (JonRowe) [inline spec/rspec/core/configuration_spec.rb:320]: Why was this changed?

File: spec/rspec/core/dsl_spec.rb [hunk 3]
```diff
@@ -6,75 +6,30 @@
...
-      end
-    end
-
-    context "when adding aliases after exposing the DSL globally" do
-      include_examples "dsl methods", :detail do
-        def changing_expose_dsl_globally
-          yield
+        def setup
           RSpec.configuration.alias_example_group_to(:detail)
         end
```
Comment (JonRowe) [inline spec/rspec/core/dsl_spec.rb:34]: Can you not just do this in a before hook here? Then you don't need the empty method call?

File: features/subject/one_liner_syntax.feature [hunk 4]
```diff
@@ -14,9 +13,7 @@ Feature: One-liner syntax
     * `is_expected` is defined simply as `expect(subject)` and is designed for
       when you are using rspec-expectations with its newer expect-based syntax.
     * `should` was designed back when rspec-expectations only had a should-based
-      syntax. However, it continues to be available and work even if the
-      `:should` syntax is disabled (since that merely removes `Object#should`
-      but this is `RSpec::Core::ExampleGroup#should`).
+      syntax.
```
Comment (JonRowe) [inline features/subject/one_liner_syntax.feature:16]: I think we should remove the `should` syntax entirely.

File: spec/rspec/core/configuration_spec.rb (hunk 2 above)
Comment (pirj) [reply spec/rspec/core/configuration_spec.rb:320]: `add_stub_and_should_receive_to` is no more.

Since the example is titled "allows rspec-mocks to be configured with a provided block", I've used one of the remaining `RSpec::Mocks::Configuration` options.

File: spec/rspec/core/configuration_spec.rb (hunk 1 above)
Comment (pirj) [reply spec/rspec/core/configuration_spec.rb:2391]: We do not monkey-patch `Module` anymore. Example group DSL is not exposed on it.

As far as I understand, in this example, we check if aliasing an example group to `my_group_method` override `ExampleGroup`'s and `Module`'s own definitions of an identically named method. This doesn't work for `Module` anymore. `Module.new.my_group_method` would return `:original`, and `< ExampleGroup` check would fail.

File: spec/rspec/core/configuration_spec.rb (hunk 2 above)
Comment (JonRowe) [reply spec/rspec/core/configuration_spec.rb:320]: Oh, silly me, I was reading this as changing from an imaginary config to an existing thing but thats not it at all. Carry on 😂

File: spec/rspec/core/configuration_spec.rb (hunk 1 above)
Comment (JonRowe) [reply spec/rspec/core/configuration_spec.rb:2391]: Got it

File: spec/rspec/core/dsl_spec.rb (hunk 3 above)
Comment (pirj) [reply spec/rspec/core/dsl_spec.rb:34]: I've started off doing exactly as you say, but it turns out `before` would run outside of `in_sub_process`, and thus litter the configuration.

It would be possible to go with `before` if `in_sub_process` could be done in `around { |example| in_sub_process { example.run } }`. But it didn't work for me just recently in [this PR](https://github.com/rspec/rspec-support/pull/452), all examples were marked as pending.
I decided to postpone the investigation on why this is the case, and use a safe approach for now.

File: features/subject/one_liner_syntax.feature [hunk 5]
```diff
@@ -9,14 +8,7 @@ Feature: One-liner syntax
...
 
-  This comes in two flavors:
-
-    * `is_expected` is defined simply as `expect(subject)` and is designed for
-      when you are using rspec-expectations with its newer expect-based syntax.
-    * `should` was designed back when rspec-expectations only had a should-based
-      syntax. However, it continues to be available and work even if the
-      `:should` syntax is disabled (since that merely removes `Object#should`
-      but this is `RSpec::Core::ExampleGroup#should`).
+  `is_expected` is a shorthand syntax for `expect(subject)`.
```
Comment (pirj) [inline features/subject/one_liner_syntax.feature:11]: @JonRowe Is this wording correct?

File: Changelog.md [hunk 6]
```diff
@@ -3,6 +3,9 @@
 Breaking Changes:
 
 * Ruby < 2.3 is no longer supported. (Phil Pirozhkov, #2787)
+* Remove `should` syntax (both monkey-patching and non-monkey-patching). (Phil Pirozhkov, #2803)
```
Comment (JonRowe) [inline Changelog.md:6]:
```suggestion
* Extract `should` syntax (including the non-monkey-patching one liner). (Phil Pirozhkov, #2803)
```

File: features/subject/one_liner_syntax.feature [hunk 7]
```diff
@@ -1,4 +1,3 @@
-@oneliner-should
 Feature: One-liner syntax
 
   RSpec supports a one-liner syntax for setting an expectation on the
```
Comment (JonRowe) [inline features/subject/one_liner_syntax.feature:3]:
```suggestion
  RSpec supports a one-liner syntax, `is_expected`,  for setting an expectation on the
```

File: features/subject/one_liner_syntax.feature [hunk 8]
```diff
@@ -9,14 +8,7 @@ Feature: One-liner syntax
   produce documentation output that does not read well or contribute to
   understanding the object you are describing.
```
Comment (JonRowe) [inline features/subject/one_liner_syntax.feature:10]:
```suggestion
  understanding the object you are describing. This syntax is a short hand for `expect(subject)`.
```

File: features/subject/one_liner_syntax.feature (hunk 5 above)
Comment (JonRowe) [inline features/subject/one_liner_syntax.feature:11]:
```suggestion
```

I'd tweak the paragraph as above (but appropriately line ended).

File: maintenance-branch [hunk 9]
```diff
@@ -1 +1 @@
-4-0-dev
\ No newline at end of file
+remove-monkey-patching
```
Comment (JonRowe) [inline maintenance-branch:1]: Is there a need for parity? whats the Pr ordering here?

**Review by JonRowe — CHANGES_REQUESTED** [review]

File: maintenance-branch (hunk 9 above)
Comment (pirj) [reply maintenance-branch:1]: Core first, then Mocks or Expectations no matter the order.

Comment (JonRowe) [reply maintenance-branch:1]: Then as this is core, this should be dropped?

Comment (pirj) [reply maintenance-branch]: Rolled it back.

**Comment by pirj** [comment]
Build failure is expected. Expectations call the removed `config.disable_monkey_patching!` method in their specs.

**Comment by JonRowe** [comment]
Ok, so this is evidently dependant on the others, which of the others can be merged without core? It seems like they should be able to be, but I haven't worked on it. When pinning repositories @pirj can you please explain what will fail without it? If they are all interdependant none of them can be merged until they are all ready for review.

**Comment by pirj** [comment]
> they are all interdependant none of them can be merged until they are all ready for review.

They are interdependent, and they are all ready for review.

**Comment by pirj** [comment]
@JonRowe 

```
/home/runner/work/rspec-core/rspec-expectations/bin/rspec

An error occurred while loading spec_helper.
Failure/Error: config.disable_monkey_patching!
```

Do you suggest to send a PR to `rspec-expectations`'s against `4-0-dev` to:
```diff
- config.disable_monkey_patching!
+ config.disable_monkey_patching! if config.respond_to?(:disable_monkey_patching!)
```
?

File: spec/rspec/core/configuration_spec.rb [hunk 10]
```diff
@@ -312,13 +312,12 @@ def mod.configuration; @config ||= Struct.new(:custom_setting).new; end
       it_behaves_like "a configurable framework adapter", :mock_with
 
       it "allows rspec-mocks to be configured with a provided block" do
-        skip unless RSpec::Mocks.configuration.respond_to?(:add_stub_and_should_receive_to)
-        mod = Module.new
+        # skip unless RSpec::Mocks.configuration.respond_to?(:add_stub_and_should_receive_to)
```
Comment (JonRowe) [inline spec/rspec/core/configuration_spec.rb:315]: This can just be removed :)

**Comment by pirj** [comment]
Green 💚

File: spec/rspec/core/configuration_spec.rb (hunk 10 above)
Comment (pirj) [reply spec/rspec/core/configuration_spec.rb:315]: Oh. 👍
